src/core/Akka/Util/Internal/AtomicState.cs

```diff
@@ -0,0 +1,186 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="AtomicState.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+//     Copyright (C) 2013-2015 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+
+namespace Akka.Util.Internal
+{
+    /// <summary>
+    /// Internal state abstraction
+    /// </summary>
+    internal abstract class AtomicState : AtomicCounterLong, IAtomicState
+    {
+        private readonly ConcurrentQueue<Action> _listeners;
+        private readonly TimeSpan _callTimeout;
+
+        protected AtomicState( TimeSpan callTimeout, long startingCount )
+            : base( startingCount )
+        {
+            _listeners = new ConcurrentQueue<Action>( );
+            _callTimeout = callTimeout;
+        }
+
+        /// <summary>
+        /// Add a listener function which is invoked on state entry
+        /// </summary>
+        /// <param name="listener">listener implementation</param>
+        public void AddListener( Action listener )
+        {
+            _listeners.Enqueue( listener );
+        }
+
+        /// <summary>
+        /// Test for whether listeners exist
+        /// </summary>
+        public bool HasListeners
+        {
+            get { return !_listeners.IsEmpty; }
+        }
+
+        /// <summary>
+        /// Notifies the listeners of the transition event via a 
+        /// </summary>
+        protected void NotifyTransitionListeners( )
+        {
+            if ( !HasListeners ) return;
+            Task
+                .Factory
+                .StartNew
```

Comment:
This is going to take a bit longer. I have two failing tests because it's swallowing the exception. Will resubmit once I resolve the failures.